### PR TITLE
fix!: refactor bindgen and update node versions

### PR DESCRIPTION
**Pull request by willemneal**
This PR updates the bindings generation for contracts to use the new and improved serialization framework `serial-as`.  The steps of the transform are now broken up into their steps.  It also opens up the use of borsh serialization.

### Initialization check required
One big change that `serial-as` makes is to the semantics of class declarations.  Assemblyscript like Typescript requires that fields that are not nullable (e.i. `string | null`) to be initialized.   Previously the bindgen code got around this requirement but calling a `defaultValue<T>()` function which would get around this check and allow non-nullables to be set as zero.

For example,

```ts

class A {
 i: i32 = 0
}

class Foo {
  a: A;
}

let f = new Foo(); // This should fail to compile because there is no initializer for `a`.

// The bindgen code would change this to 

class Foo {
  a: A = defaultValue<A>(); // Does `changetype<A>(0)` <-- unsafe `null`
}

// Can still have empty initializer if nullable

class Foo {
  a: A | null;
}
// Now it is clear that the field might be null whereas before it was assumed not be.

```
This original transform was there for backwards compatibility to previous versions of AS which didn't have the initialization check.

However, this led to encouraging bad and unsafe practices and thus now all fields are required to be initialized.

### Node versions

This PR also removes compatibility to node versions `12` and `13`.  The former does not support optional chaining `x?.foo()`, and the later is dropped by newer versions of jest.

### CODE REVIEW

File: bindgen/assembly/index.ts [hunk 1]
```diff
@@ -98,404 +92,58 @@ function getInput(): JSON.Obj {
...
-}
+// @global
+// function __defaultValue<T>(): T {
+//   if (isInteger<T>() || isFloat<T>()) {
+//     // @ts-ignore
+//     return <T>0;
+//   }
+//   if (isString<T>()) return "";
+//   return changetype<T>(0);
+// }
```
Comment (chadoh) [inline bindgen/assembly/index.ts:146]: can rm comment?

File: bindgen/package.json [hunk 2]
```diff
@@ -35,6 +35,10 @@
     "assembly/index.ts"
   ],
   "devDependencies": {
-    "near-sdk-core": "^3.2.1"
+    "near-sdk-core": "^3.2.0"
```
Comment (chadoh) [inline bindgen/package.json:38]: reverting to an older version of sdk-core?

File: bindgen/src/JSONBuilder.ts [hunk 3]
```diff
@@ -1,325 +1,56 @@
...
+//       );
+//     }
+//   }
+
+//   static visit(sources: Source[]): Source[] {
+//     const builder = new JSONBindingsBuilder();
+//     builder.visit(sources);
+//     return sources;
+//   }
+// }
```
Comment (chadoh) [inline bindgen/src/JSONBuilder.ts:56]: can rm comments (whole file)?

File: bindgen/src/transformer.ts [hunk 4]
```diff
@@ -1,59 +1,35 @@
-import { Transform, Parser, Source, Module, SourceKind } from "visitor-as/as";
-import { JSONBindingsBuilder, isEntry } from "./JSONBuilder";
-import { TypeChecker } from "./typeChecker";
+import { Transform, Parser, Program, Module, SourceKind } from "visitor-as/as";
+// import { JSONBindingsBuilder } from "./JSONBuilder";
```
Comment (chadoh) [inline bindgen/src/transformer.ts:2]: can rm comment?

File: sdk-core/assembly/index.ts [hunk 5]
```diff
@@ -9,5 +9,8 @@ export * from "./logging";
 export * from "./math";
 export * from "./promise";
 export * from "./datetime";
+import * as JSON from "@serial-as/json";
+export { JSON };
+export * from "@serial-as/borsh";
```
Comment (chadoh) [inline sdk-core/assembly/index.ts:14]: What are the implications of this change? I didn't see it mentioned in any updated docs or in the PR description (do we use the PR description to create release notes? might be useful)

File: sdk/assembly/__tests__/bignum.spec.ts [hunk 6]
```diff
@@ -1,18 +1,22 @@
 import { u128 } from "..";
 
+// TODO: findout why the commented lines below fail
```
Comment (chadoh) [inline sdk/assembly/__tests__/bignum.spec.ts:3]: I assume this TODO comment needs to be addressed prior to merge?

**Review by chadoh — CHANGES_REQUESTED** [review]
See some notes below. I think there's at least some minor cleanup & issues to address prior to merging. I also don't see any doc updates, and some changes aren't mentioned in the PR description (which seems like a good place to "stage" release notes, and we'll definitely want this stuff mentioned in the release notes with the next major version).

File: bindgen/package.json (hunk 2 above)
Comment (willemneal) [reply bindgen/package.json:38]:
```suggestion
    "near-sdk-core": "^3.2.1"
```

File: sdk-core/assembly/index.ts (hunk 5 above)
Comment (willemneal) [reply sdk-core/assembly/index.ts:14]: In the future branch this isn't needed because we have a `serializer.ts` which exports a simplified serializer.  So I think I'll add that to this PR.  And add to PR description.  I also think we should make a CHANGELOG.md to formalize this.

File: sdk/assembly/__tests__/bignum.spec.ts (hunk 6 above)
Comment (willemneal) [reply sdk/assembly/__tests__/bignum.spec.ts:3]: Yeah good point.  bignum didn't fail in any of the other tests so I saw this as a future work to understand this edge case.
